diffkemp/simpll/PatternSet.cpp

```diff
@@ -0,0 +1,420 @@
+        PatternSet::DefaultPrefix + PatternSet::PrefixR;
+/// Name for the function defining output instuction mapping.
+const std::string PatternSet::OutputMappingFunName =
+        PatternSet::DefaultPrefix + "mapping";
+/// Name for pattern metadata nodes.
+const std::string PatternSet::MetadataName =
+        PatternSet::DefaultPrefix + "pattern";
+
+/// Create a new pattern set based on the given configuration.
+PatternSet::PatternSet(std::string ConfigPath) {
+    if (ConfigPath.empty()) {
+        return;
+    }
+
+    // If a pattern is used as a configuration file, only load the pattern.
+    if (extension(ConfigPath) == ".ll") {
+        addPattern(ConfigPath);
+    } else {
+        loadConfig(ConfigPath);
+    }
+}
+
+/// Retrives pattern metadata attached to the given instruction.
+Optional<PatternMetadata>
+        PatternSet::getPatternMetadata(const Instruction &Inst) const {
+    auto InstMetadata = Inst.getMetadata(MetadataName);
+    if (!InstMetadata) {
+        return Optional<PatternMetadata>{};
+    }
+
+    int OperandIndex = 0;
+    PatternMetadata Metadata;
+    while (OperandIndex < InstMetadata->getNumOperands()) {
+        /// Parse the current pattern metadata operand as well as the operands
+        /// that depend on it.
+        if (auto Type = dyn_cast<MDString>(
+                    InstMetadata->getOperand(OperandIndex).get())) {
+            auto TypeName = Type->getString();
+            if (TypeName == "pattern-start") {
+                // pattern-start metadata: string type.
+                Metadata.PatternStart = true;
+            } else if (TypeName == "pattern-end") {
+                // pattern-end metadata: string type.
+                Metadata.PatternEnd = true;
+            } else if (TypeName == "group-start") {
+                // group-start metadata: string type.
+                Metadata.GroupStart = true;
+            } else if (TypeName == "group-end") {
+                // group-end metadata: string type.
+                Metadata.GroupEnd = true;
+            } else if (TypeName == "disable-name-comparison") {
+                // disable-name-comparison metadata: string type.
+                Metadata.DisableNameComparison = true;
+            } else if (TypeName == "not-an-input") {
+                // not-an-input metadata: string type.
+                Metadata.NotAnInput = true;
+            } else {
+                DEBUG_WITH_TYPE(DEBUG_SIMPLL,
+                                dbgs() << getDebugIndent()
+                                       << "Invalid metadata type " << TypeName
+                                       << " in node " << *InstMetadata
+                                       << ".\n");
+                return Optional<PatternMetadata>{};
+            }
+            // Shift the operand offset accordingly.
+            OperandIndex += PatternMetadata::MetadataOffsets.lookup(TypeName);
+        } else {
+            return Optional<PatternMetadata>{};
```

Comment:
Doesn't LLVM's `Optional` have something like `std::nullopt`?

Reply:
I may be missing something, but no, I do not see anything like that.

Reply:
There's `llvm::None` which should be usable for this case.
https://llvm.org/doxygen/None_8h.html